File: cpp/198.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

class Solution {
public:
  int rob(vector<int> &nums) {
    int prepre = 0, pre = 0, cur = 0;
    int n = nums.size();
    for (int i = 0; i < n; i++) {
      cur = max(nums[i] + prepre, pre);
      prepre = pre;
      pre = cur;
    }
    return cur;
  }
};

int main(){
  vector<int> v{1,1,1,1,1};
  cout << Solution().rob(v);
}

File: python/198.py
xrange = range

class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        l = [0] * n
        for i in xrange(n):
            l[i] = max(nums[i] + (0 if i <= 1 else l[i-2]),
                       0 if i == 0 else l[i-1])
        return l[-1]

print(Solution().rob([1,1,1,1,1]))
